db/migrate/20180510184222_update_foreign_key_constraints.rb

```diff
@@ -0,0 +1,311 @@
+class UpdateForeignKeyConstraints < ActiveRecord::Migration
+  def self.up
+	execute %q{
+
+/* These constraints are being updated to use the "ON UPDATE CASCADE" clause so
+that rows having misallocated id numbers can be updated to use ids in the
+correct range without violating foreign key constraints.  A few constraints are
+new. */
+
+ALTER TABLE "posteriors_ensembles" DROP CONSTRAINT "fk_posteriors_ensembles_ensembles_1";
+ALTER TABLE "posteriors_ensembles" ADD CONSTRAINT "fk_posteriors_ensembles_ensembles_1" FOREIGN KEY ("ensemble_id") REFERENCES "ensembles" ("id") ON DELETE CASCADE ON UPDATE CASCADE;
+
+ALTER TABLE "runs" DROP CONSTRAINT "fk_runs_ensembles_1";
+ALTER TABLE "runs" ADD CONSTRAINT "fk_runs_ensembles_1" FOREIGN KEY ("ensemble_id") REFERENCES "ensembles" ("id") ON UPDATE CASCADE;
+
+ALTER TABLE "inputs" DROP CONSTRAINT "fk_inputs_inputs_1";
+ALTER TABLE "inputs" ADD CONSTRAINT "fk_inputs_inputs_1" FOREIGN KEY ("parent_id") REFERENCES "inputs" ("id") ON UPDATE CASCADE         not valid;
```

Comment:
how did you decide which tables to use `not valid` to allow the constraint to be set even on tables where the constraint condition is not met? Is it only based on the tables in bety_mepp that violate the constraints?

Reply:
Yes.  Since the motivation for this migration was fixing `bety_mepp`, I tried it out on a copy of `bety_mepp` and applied `NOT VALID` as needed.  (There may have also been cases where the constraint I was adding `ON UPDATE CASCADE` to was already marked `NOT VALID` and I didn't see fit to change it.)

_Most_ of the problems in `bety_mepp` arise when a table is imported from another database (usually `ebi_production` on `ebi-forecast`) but the tables it refers to are not.  Foreign-key constraints are lifted during imports and they aren't enforced for the imported data (though they are put back in place for "new" data).  On the other hand, a uniqueness constraint violation can cause a particular table not to get imported even if other tables are.

There are a few cases where the constraints would be violated even if all data were successfully imported from all machines.